README.md
CoastCR <img src="man/figures/logo.png" align="right" alt="" width="140" />
=========================================================
# `CoastCR`: Coastal Change using R

<!-- badges: start -->
[![R-CMD-check](https://github.com/alejandro-gomez/CoastCR/workflows/R-CMD-check/badge.svg)](https://github.com/alejandro-gomez/CoastCR/actions)
[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
<!-- badges: end -->
  
## Overview

**CoastCR** is an open source package for computing coastline time series and trend analysis using the baseline and transect approach. This tool is a part of ODSAS (Open Digital Shoreline Analysis System v1.0) which uses both CoastCR and SAGA GIS [1, 2]. CoastCR main input is the points representing the intersections along transects perpendicular to a baseline, of a set of date-stamped coastlines (obtained during mapping stage using SAGA GIS) and calculates the standard metrics of coastline change and rate of change similarly to the metrics produced using Digital Shoreline Analysis System (DSAS) [3, 4]. Package website: https://alejandro-gomez.github.io/CoastCR/

## Current version

**CoastCR 1.2.0**.

This version incorporates a change in the Date format. Currently, the user can select more than one shoreline for each day. For this, the user should be introduced to the table with the dates and associated uncertainty in a column named "Day" in format (YYYYY-mm-dd). If the information about the acquisition hour is available, the user should introduce a column called "Hour" in format (HH:MM:SS). Now the column names should be "Day", "Hour" and "Uncertainty". If the data does not include the hour information, remove the column to avoid possible errors.


----

Contents:

* [What is this package used for?](#what-is-this-package-used-for)
* [How it works](#how-it-works)
  * [Installation](#installation)
  * [Example](#example)
* [Package citation](#package-citation)
* [Authors](#authors)
* [Contact](#contact)
* [References](#references)
* [Citations](#citations)

----

## What is this package used for?
Many researchers and practitioners interested on Shoreline Change Analysis often separate the mapping stages (mostly done within a GIS) from the time series and trend analyses, which can be undertaken in a programming environment [1]. This package is intended to be used for the time series and trend analyses using R programming environment which is an open-source software.

The main functions implemented in, `CoastCR` are used to:

-    Compute the **coastline change** using the most common change metrics used on transects to a baseline method. Two main functions are included in the current version to filter and estimate the parameters:

     - ***Baseline filter***: This function is part of the pre-processing stages and is used to ensure that for each baseline-transect, only one crossing with each date-stamped coastline is used. If for a given transect and date-stamped coastline, more than one crossing is found (e.g., as might happen for very wiggly coastlines), the filter selects the crossing that is closest to the baseline and neglects the others.
  
     - ***Coast rates***: This function estimates the main five key measures for each individual transect and also produce aggregates for all transects identically as within the DSAS tool [2]. This five metrics are:
     
        - **NSM**: Net Shoreline Movement (m).
        - **EPR**: End Point Rate (m yr<sup>-1</sup>).
        - **SCE**: Shoreline Change Envelope (m).
        - **LRR**: Linear Regression Rate (m yr<sup>-1</sup>).
        - **WLR**: Weighted Linear Regression Rate (m yr<sup>-1</sup>).

&nbsp;

- ***Coastline Variations*** is the main function that runs the aforementioned functions. First, filters the intersection points and then calculates the key measures of change.
        
## How it works

### Installation

``` r
# To install the latest version from Github:
# install.packages("remotes")
remotes::install_github("alejandro-gomez/CoastCR")
```

### Example

``` r
# Load libraries
library(sf)
library(CoastCR)

# Intersections shapefile
shp <- st_read(system.file("./extdata/intersect.shp", package = "CoastCR"))

# Normal lines shapefile
normals <- st_read(system.file("./extdata/normals.shp", package = "CoastCR"))

# Table with dates and associated uncertainty
table <- read.csv(system.file("./extdata/table_coastlines.csv", package = "CoastCR"))

# Define baseline position. Offshore = OFF; Onshore = ON; Mixed = MIX.
position = "OFF"

# Define outputs names
out_points <- "./int_filter.shp"
out_name <- "./normals_rates.shp"


coast_var(shp, normals, table, position, out_points, out_name)
```

## Package citation

Using CoastCR for research publication?  Please **cite it**! I am an early career scientist and every citation matters.

***Gómez-Pazo, A., Payo, A., Paz-Delgado, M.V., Delgadillo-Calzadilla, M.A.*** (*2022*). *Open Digital Shoreline Analysis System: ODSAS v1.0*. Journal of Marine Science and Engineering, 10, 26. DOI: https://doi.org/10.3390/jmse10010026

## Authors

Alejandro Gómez-Pazo, Andres Payo and M. Victoria Paz-Delgado

Contributors: M.A. Delgadillo-Calzadilla

## Contact

Alejandro Gómez-Pazo: a.gomez@usc.es

## References

[1] Gómez-Pazo, A., Payo, A., Paz-Delgado, M.V., Delgadillo-Calzadilla, M.A., 2022. Open Digital Shoreline Analysis System: ODSAS v1.0. Journal of Marine Science and Engineering, 10, 26. DOI: https://doi.org/10.3390/jmse10010026

[2] Paz-Delgado, MV., Payo, A., Gómez-Pazo, A., Beck, A-L., Savastano, S., 2022. Shoreline Change from Optical and Sar Satellite Imagery at Macro-Tidal Estuarine, Cliffed Open-Coast and Gravel Pocket-Beach Environments. Journal of Marine Science and Engineering, 10(5), 561. DOI: https://doi.org/10.3390/jmse10050561

[3] Himmelstoss, E.A., Henderson, R.E., Kratzmann, M.G., Farris, A.S., 2018. Digital Shoreline Analysis System (DSAS) version 5.0 user guide. US Geological Survey Open File Report 2018-1179, 110 pp

[4] Burningham, H.; Fernandez-Nunez, M. 19 - shoreline change analysis. In Sandy beach morphodynamics, Jackson, D.W.T.; Short, A.D., Eds. Elsevier: 2020; pp 439-460.

## Citations

Paz-Delgado, MV., Payo, A., Gómez-Pazo, A., Beck, A-L., Savastano, S., 2022. Shoreline Change from Optical and Sar Satellite Imagery at Macro-Tidal Estuarine, Cliffed Open-Coast and Gravel Pocket-Beach Environments. Journal of Marine Science and Engineering, 10(5), 561. DOI: https://doi.org/10.3390/jmse10050561

Ankrah, J.; Monteiro, A.; Madureira, H., 2022. Bibliometric Analysis of Data Sources and Tools for Shoreline Change Analysis and Detection. Sustainability, 14(9), 4895. DOI: https://doi.org/10.3390/su14094895

Vallarino Castillo, R.; Negro Valdecantos, V.; Moreno Blasco, L., 2022. Shoreline Change Analysis Using Historical Multispectral Landsat Images of the Pacific Coast of Panama. Journal of Marine Science and Engineering, 10(12), p.1801. DOI: https://doi.org/10.3390/jmse10121801

Tsiakos, C-A.D.; Chalkias, C., 2023. Use of Machine Learning and Remote Sensing Techniques for Shoreline Monitoring: A Review of Recent Literature. Applied Sciences 13 (5), 3268. DOI: https://doi.org/10.3390/app13053268

Hossen, M.F.; Sultana, N., 2023. Shoreline change detection using DSAS technique: Case of Saint Martin Island, Bangladesh. Remote Sensing Applications: Society and Environment, 30, p.100943. DOI: https://doi.org/10.1016/j.rsase.2023.100943

Gómez-Pazo, A., 2023. El empleo de metodologías de código abierto para las investigaciones costeras: comparativa de las técnicas de detección de cambios. Boletín de la Asociación de Geógrafos Españoles, (96). DOI: https://doi.org/10.21138/bage.3318


NEWS.md
# CoastCR 1.2.0

* This version incorporates a change in the Date format. Currently, the user can select more than one shoreline for each day. For this, the user should be introduced to the table with the dates and associated uncertainty in a column named "Day" in format (YYYYY-mm-dd). If the information about the acquisition hour is available, the user should introduce a column called "Hour" in format (HH:MM:SS). Now the column names should be "Day", "Hour" and "Uncertainty". If the data does not include the hour information, remove the column to avoid possible errors.

# CoastCR 1.1.0

* The stats summary are saving as csv file and include quantiles information.

# CoastCR 1.0.0

* Original version. Results summary save as png format file.


tests/testthat.R
library(testthat)
library(CoastCR)

test_check("CoastCR")


tests/testthat/test-coast_var.R
test_that("coast_rates", {
  testthat::skip_on_cran()

  test_dir <- tempdir()

  setwd(test_dir)

  dist <- st_read(system.file("./extdata/normals_coast_rate.shp", package = "CoastCR"))

  shp <- st_read(system.file("./extdata/intersect.shp", package = "CoastCR"))

  normals <- st_read(system.file("./extdata/normals.shp", package = "CoastCR"))

  table <- read.csv(system.file("./extdata/table_coastlines.csv", package = "CoastCR"))

  position = "OFF"

  out_points <- "./int_filter.shp"
  out_name <- "./normals_rates.shp"

  coast_var(shp, normals, table, position, out_points, out_name)
  res <- st_read("./normals_rates.shp")

  expect_equal(res, dist, tolerance = 0.0009)

  unlink(test_dir, recursive = TRUE)

})


tests/testthat/test-coast_rates.R
test_that("coast_rates", {
  testthat::skip_on_cran()

  setwd(tempdir())

  normals <- st_read(system.file("./extdata/normals.shp", package = "CoastCR"))

  table <- read.csv(system.file("./extdata/table_coastlines.csv", package = "CoastCR"))

  inter_dist <- st_read(system.file("./extdata/dist.shp", package = "CoastCR"))

  dist <- st_read(system.file("./extdata/normals_coast_rate.shp", package = "CoastCR"))

  out_name <- "./normals_coast_rates.shp"

  coast_rates(inter_dist, normals, table, out_name)

  res <- st_read("./normals_coast_rates.shp")

  expect_equal(res, dist, tolerance = 0.0009)

})


R/coast_var.R
#' Filter intersections and estimate variation rates
#'
#' \code{coast_var} allows to filter the original intersections and select only the closest intersection between each transect and shoreline.
#'
#' @param shp Shapefile. A point format shapefile with the intersections between each transect and each shoreline.
#' @param normals Shapefile. A polyline format shapefile with all normal lines (transects) included in the study site.
#' @param table CSV. A table with the information about the coastlines dates in a column named "Day" in format (YYYYY-mm-dd). If the information about the acquisition hour is available, the user should introduce a column called "Hour" in format (HH:MM:SS) and the associated uncertainty to each coastline in meters. The column names should be "Day", "Hour" and "Uncertainty".
#' @param position character. Default \code{"MIX"}. \code{position = "MIX"} is recommended for use a coastline as baseline. The baseline position in relation to the coastlines. There are three options: "MIX" for intermediate baselines, "OFF" for offshore baselines, and "ON" for onshore baselines.
#' @param out_points Integer. Output name for the filter intersection points shapefile.
#' @param out_name Integer. Output name for the resulting shapefile with the rates of each transect.
#'
#' @details
#' The \code{coast_var} is the main function included in this \code{CoastCR} version, as part of ODSAS (Gomez-Pazo et al., \emph{2022}).
#' This function runs \code{baseline_filter} and \code{coast_rates} to doing the entire coastal change process.
#' The first step filters the intersection points, to remove the duplicate points
#' and selects the crossing that is closest to the baseline and neglects the others.
#' The second step estimates the main coastal variation rates explained in detail in \code{CoastCR} documentation.
#'
#' @return \itemize{
#'    \item{A shapefile with the filtered intersection points}
#'    \item{A polyline shapefile with all rates associated to each transect.}
#'    \item{A table in png format with the central tendency stats for each parameter.}
#' }
#'
#'
#' @examples
#'
#' #Load libraries
#' library(sf)
#' library(CoastCR)
#'
#' #Intersections shapefile
#' shp <- st_read(system.file("extdata/intersect.shp", package = "CoastCR"))
#'
#' #Normal lines shapefile
#' normals <- st_read(system.file("extdata/normals.shp", package = "CoastCR"))
#'
#' #Table with dates and associated uncertainty
#' table <- read.csv(system.file("extdata/table_coastlines.csv", package = "CoastCR"))
#'
#' #Define baseline position. Offshore = OFF; Onshore = ON; Mixed = MIX.
#' position = "OFF"
#'
#' #Define outputs names
#' out_points <- tempfile("int_filter", fileext = ".shp")
#' out_name <- tempfile("normals_rates", fileext = ".shp")
#'
#' coast_var(shp, normals, table, position, out_points, out_name)
#'
#'
#' @references {
#' Gomez-Pazo, A., Payo, A., Paz-Delgado, M.V., Delgadillo-Calzadilla, M.A. (2022)
#' \emph{Open Digital Shoreline Analysis System: ODSAS v1.0}
#' Journal of Marine Science and Engineering, 10, 26.
#'}
#' @seealso  \code{\link{baseline_filter}}; \code{\link{coast_rates}}
#'
#' @export

coast_var <- function(shp, normals, table, position = "MIX",
                      out_points, out_name) {
  baseline_filter(shp, position, out_points)

  int_d <- st_read(out_points)

  coast_rates(int_d, normals, table, out_name)

}


R/baseline_filter.R
#' Filter the intersections between transects and shorelines
#'
#' \code{baseline_filter} allows to filter the original intersections and select only the closest intersection between each transect and coastline.
#'
#' @param shp Shapefile. A point format shapefile with the intersections between each transect and each shoreline.
#' @param position character. Default \code{"MIX"}. \code{position = "MIX"} is recommended for use a coastline as baseline. The baseline position in relation to the coastlines. There are three options: "MIX" for intermediate baselines, "OFF" for offshore baselines, and "ON" for onshore baselines.
#' @param out_points Integer. Output name for the filter intersection points shapefile.
#'
#' @details
#' This function is part of the pre-processing stages and is used to ensure that for each baseline-transect,
#' only one crossing with each date-stamped coastline is used.
#' The filter selects the crossing that is closest to the baseline and neglects the others (Gomez-Pazo et al., \emph{2022}).
#'
#' @return \itemize{
#'    \item{A shapefile with the filtered intersection points, only one point for each transect/shoreline pair.}
#'}
#'
#'@references {
#' Gomez-Pazo, A., Payo, A., Paz-Delgado, M.V., Delgadillo-Calzadilla, M.A. (2022)
#' \emph{Open Digital Shoreline Analysis System: ODSAS v1.0}
#' Journal of Marine Science and Engineering, 10, 26.
#'}
#' @seealso  \code{\link{coast_var}}; \code{\link{coast_rates}}
#'
#' @import sf dplyr tidyverse tidyr stringr
#'
#' @examples
#'
#' #Load libraries
#' library(sf)
#' library(CoastCR)
#'
#' #Intersections shapefile
#' shp <- st_read(system.file("extdata/intersect.shp", package = "CoastCR"))
#'
#' #Define baseline position. Offshore = OFF; Onshore = ON; Mixed = MIX.
#' position = "OFF"
#'
#' #Define output name
#' out_points <- tempfile("inters_filter", fileext = ".shp")
#'
#' baseline_filter(shp, position, out_points)
#'
#' @export

baseline_filter <- function(shp, position = "MIX", out_points) {
  if (position == "OFF"){
    shp2 <- shp %>%
      dplyr::filter(Distance < 0) %>%
      unite(ID_02, c("ID_Profile", "ID_Coast"))%>%
      group_by(ID_02) %>%
      dplyr::filter(Distance == max(Distance)) %>%
      dplyr::distinct(ID_02, .keep_all = TRUE) %>%
      mutate(Normal = sub("_.*","", ID_02),
             Coast = sub(".*_","", ID_02))
    st_write(shp2, out_points)

  } else if (position == "ON"){
    shp2 <- shp %>%
      dplyr::filter(Distance > 0) %>%
      unite(ID_02, c("ID_Profile", "ID_Coast"))%>%
      group_by(ID_02) %>%
      dplyr::filter(Distance == min(Distance)) %>%
      dplyr::distinct(ID_02, .keep_all = TRUE) %>%
      mutate(Normal = sub("_.*","", ID_02),
             Coast = sub(".*_","", ID_02))
    st_write(shp2, out_points)

  } else if (position == "MIX"){
    shp2 <- shp %>%
      mutate(abs = abs(Distance)) %>%
      unite(ID_02, c("ID_Profile", "ID_Coast"))%>%
      group_by(ID_02) %>%
      dplyr::filter(abs == min(abs))%>%
      dplyr::distinct(ID_02, .keep_all = TRUE) %>%
      mutate(Normal = sub("_.*","", ID_02),
             Coast = sub(".*_","", ID_02))
    st_write(shp2, out_points)

  } else{
    stop("Baseline position or input shapefile are not correct")
  }
}



R/coast_rates.R
#' Calculate the coastal variation rates
#'
#' \code{coast_rates} estimate the main parameters about the coastal variations, as NSM, LRR or WLR.
#'
#' @param inter_dist Shapefile. A point format shapefile with the intersections between each transect and the shoreline.
#' @param normals Shapefile. A polyline format shapefile with all normal lines (transects) included in the study site.
#' @param table CSV. A table with the information about the coastlines dates in a column named "Day" in format (YYYYY-mm-dd). If the information about the acquisition hour is available, the user should introduce a column called "Hour" in format (HH:MM:SS) and the associated uncertainty to each coastline in meters. The column names should be "Day", "Hour" and "Uncertainty".
#' @param out_name Integer. Output name for the resulting shapefile with the rates of each transect.
#'
#' @details
#' This function is part of ODSAS method (Gomez-Pazo et al., \emph{2022}) and estimates the main five key measures for each individual transect and also produce aggregates
#' for all transects identically as within the DSAS tool (Himmelstoss et al., 2018).
#' The rates and their units are explained in detail in \code{CoastCR} documentation.
#'
#' @return \itemize{
#'    \item{A polyline shapefile with all rates associated to each transect.}
#'    \item{A csv file with the central tendency stats and quantiles .25, .75 and .90 for each parameter.}
#' }
#'
#' @references {
#' Gomez-Pazo, A., Payo, A., Paz-Delgado, M.V., Delgadillo-Calzadilla, M.A. (2022)
#' \emph{Open Digital Shoreline Analysis System: ODSAS v1.0}
#' Journal of Marine Science and Engineering, 10, 26.
#'
#' Himmelstoss, E.A., Henderson, R.E., Kratzmann, M.G., Farris, A.S. (2018)
#' \emph{Digital Shoreline Analysis System (DSAS) version 5.0 user guide}
#' US Geological Survey Open File Report 2018-1179, 110 pp
#'}
#' @seealso  \code{\link{baseline_filter}}; \code{\link{coast_var}}
#'
#' @import qwraps2
#'
#' @importFrom stats dist
#' @importFrom stats na.omit
#' @importFrom utils write.csv
#' @importFrom psych describe
#'
#' @examples
#' library(sf)
#' library(CoastCR)
#'
#' #Normal lines shapefile
#' normals <- st_read(system.file("extdata/normals.shp", package = "CoastCR"))
#'
#' #Table with dates and associated uncertainty
#' table <- read.csv(system.file("extdata/table_coastlines.csv", package = "CoastCR"))
#'
#' #Filtered intersections shapefile
#' inter_dist <- st_read(system.file("extdata/dist.shp", package = "CoastCR"))
#'
#' #Define output name
#' out_name <- tempfile("normals_coast_rates", fileext = ".shp")
#'
#' coast_rates(inter_dist, normals, table, out_name)
#'
#'
#' @export

coast_rates <- function(inter_dist, normals, table, out_name) {
  if ("Hour" %in% colnames(table)) {
    table$Date <- as.POSIXlt(paste(table$Day, table$Hour, sep = ""), format = "%Y-%m-%d %H:%M:%S")
  } else {
    table$Date <- as.POSIXlt(table$Day, format = "%Y-%m-%d")
  }

  list1 <- NULL
  for (i in 1:nrow(table)){
    name_d <- paste("D", table$Date[i], sep = "_")
    name_t <- paste("Y", table$Date[i], sep = "_")
    a <- inter_dist %>%
      st_drop_geometry() %>%
      mutate(dist = ifelse(Coast == i-1, Distance, NA)) %>%
      group_by(Normal) %>%
      dplyr::filter(!is.na(dist)) %>%
      dplyr::select(Normal, dist)
    names(a) <- c("Normal", paste("D", table$Date[i], sep = "_"))
    assign(paste("Y", table$Date[i], sep = "_"), a)
    list1[[i]] <- a
    yrs <- Reduce(function(...) merge(..., by="Normal", all = TRUE), list1)
  }
  mm_Dates <- c(paste("D", min(table$Date), sep = "_"),
                paste("D", max(table$Date), sep = "_"))
  SCE_val <- NULL
  normals2 <- normals %>%
    dplyr::select(1)%>%
    merge(yrs, by = "Normal", all.x = TRUE)%>%
    mutate(NSM = (get(mm_Dates[2]) - get(mm_Dates[1])),
           EPR = NSM / (as.numeric(abs(difftime(min(table$Date),
                                                max(table$Date), units = "days")/365))),
           EPRunc = (sqrt((as.numeric(
             table$Uncertainty[table$Date==max(table$Date)])^2) +
               (as.numeric(
                 table$Uncertainty[table$Date==min(table$Date)])^2)))/
             (as.numeric(abs(difftime(min(table$Date), max(table$Date),
                                      units = "days")/365)))) %>%
    dplyr::select(Normal, NSM, EPR, EPRunc, everything())

  list2 <- NULL
  SCE_comp <- normals2 %>%
    st_drop_geometry() %>%
    mutate(SCE_max = do.call(pmax, c(select(., 5:ncol(.)), na.rm = TRUE)),
           SCE_min = do.call(pmin, c(select(., 5:ncol(.)), na.rm = TRUE)),
           SCE = abs(SCE_max - SCE_min)) %>%
    dplyr::select(Normal, SCE)
  normals3 <- merge(normals2, SCE_comp, by="Normal", all = TRUE) %>%
    dplyr::select(Normal, NSM, EPR, EPRunc, SCE, everything())

  list3 <- NULL
  ancient <- paste("D", min(table$Date), sep = "_")
  for (z in 1:nrow(normals3)){
    St_P <- NULL
    for (k in 1:nrow(table)){
      if (ancient == paste("D", table$Date[k], sep = "_")){
        Dist_c <- normals[z,] %>%
          st_drop_geometry() %>%
          mutate(Date_C = as.numeric(0),
                 Distance_C = as.numeric(0),
                 w_C = 1/(as.numeric(table$Uncertainty[k])^2)) %>%
          dplyr::select(Normal, Date_C, Distance_C, w_C)
        assign(paste("Dist_n", table$Date[k], sep = "_"), Dist_c)
        list3 <- Dist_c
      } else if (ancient != paste("D", table$Date[k], sep = "_")){
        Dist_c <- normals3[z,] %>%
          st_drop_geometry() %>%
          mutate(Date_C = as.numeric(abs(difftime(min(table$Date),
                                                  table$Date[k], units = "days")/365)),
                 Distance_C = as.numeric(get(paste("D", table$Date[k],
                                                   sep = "_")) -
                                           get(paste("D", min(table$Date), sep = "_"))),
                 w_C = 1/(as.numeric(table$Uncertainty[k])^2)) %>%
          dplyr::select(Normal, Date_C, Distance_C, w_C)
        assign(paste("Dist_n", table$Date[k], sep = "_"), Dist_c)
        list3 <- Dist_c
      } else{
        stop("Dates information is not correct, check date format.")
      }
      St_P <- rbind(St_P, na.omit(list3))
    }
    mean_Date <- mean(St_P$Date_C, na.rm = TRUE)
    mean_Dist <- mean(St_P$Distance_C, na.rm = TRUE)
    SS <- 0
    SCP <- 0
    totalSS <- 0
    for (x in 1:nrow(St_P)){
      SS <- (SS + (St_P$Date_C[x]- mean_Date)^2)
      SCP <- (SCP + (St_P$Date_C[x]- mean_Date)*(St_P$Distance_C[x]- mean_Dist))
      totalSS <- totalSS + ((St_P$Distance_C[x]- mean_Dist)^2)
    }
    b <- SCP/SS
    LRR <- b
    normals3$LRR[z] <- LRR
    r2 <- b * SCP
    residualsSS <- totalSS - (b * SCP)
    LR2 <- (b * SCP) / totalSS
    normals3$LR2[z] <- LR2
    sumW <- 0
    sumWX <- 0
    sumWY <- 0
    sumWXY <- 0
    sumWYY <- 0
    sumWXX <- 0
    for (v in 1:nrow(St_P)){
      sumW <- sumW + St_P$w_C[v]
      sumWY <- sumWY + (St_P$w_C[v] * St_P$Distance_C[v])
      sumWX <- sumWX + (St_P$w_C[v] * St_P$Date_C[v])
      sumWXX <- sumWXX + (St_P$w_C[v] * St_P$Date_C[v] * St_P$Date_C[v])
      sumWYY <- sumWYY + (St_P$w_C[v] * St_P$Distance_C[v] * St_P$Distance_C[v])
      sumWXY <- sumWXY + (St_P$w_C[v] * St_P$Date_C[v] * St_P$Distance_C[v])
      SS_w <- sumWXX - (sumWX * sumWX/sumW)
      SCP_w <- sumWXY - (sumWX * sumWY/sumW)
      totalSS_w <- sumWYY - (sumWY * sumWY / sumW)
    }
    b_w <- SCP_w/SS_w
    WLR <- b_w
    normals3$WLR[z] <- WLR
    regressionSS_W <- b_w * SCP_w
    residualSS_W <- totalSS_w - regressionSS_W
    WR2 <- regressionSS_W / totalSS_w
    normals3$WR2[z] <- WR2
  }
  normals4 <- normals3 %>%
    dplyr::select("Normal", "NSM", "EPR", "EPRunc",
                  "SCE", "LRR", "LR2", "WLR", "WR2")
  st_write(normals4, out_name)

  n4 <- normals4%>%
    st_drop_geometry()
  summary_data <- describe(n4[ ,c("NSM", "EPR", "EPRunc", "SCE", "LRR", "WLR")],
                           quant = c(.25,.75,.90)) %>%
    dplyr::select("n", "mean", "sd", "median", "min", "max", "range",
                  "Q0.25", "Q0.75", "Q0.9")
  names(summary_data) <- c("n", "Mean", "SD", "Median", "min", "Max", "Range",
                           "Quantile .25", "Quantile .75", "Quantile .9")
  write.csv(summary_data, paste0(gsub(".shp*$","_summary",out_name),".csv"))
}
